### PR TITLE
[NEEDS TESTING] Axes tcomms

### PR DESCRIPTION
**Pull request by Kelenius**
Replaces with a magic box that makes radios work when on. Works regardless of z-level.
Removes icon of the radio used because it's a hassle to maintain.
PDA spamfilter is now global for all messaging servers.

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
You have a lot of "src." in your code that are unnecessary; remember, src. is already implied.

**Comment by Zuhayr** [comment]
Thanks for your input.

File: code/game/objects/items/devices/PDA/PDA.dm [hunk 1]
```diff
@@ -978,17 +984,17 @@ var/global/list/obj/item/device/pda/PDAs = list()
...
+		if(!P.has_connection()) // Does our recipient have a broadcaster on their level?
 			U << "ERROR: Cannot reach recipient."
 			return
-		var/send_result = reception.message_server.send_pda_message("[P.owner]","[owner]","[t]")
-		if (send_result)
-			U << "ERROR: Messaging server rejected your message. Reason: contains '[send_result]'."
-			return
+
+		for(var/obj/machinery/message_server/M in message_servers)
+			var/send_result = M.send_pda_message("[P.owner]","[owner]","[t]") // We need to make logs in all servers
```
Comment (PsiOmegaDelta) [inline code/game/objects/items/devices/PDA/PDA.dm]: Does this mean that you'll get one PDA message for each active message server?

Comment (Kelenius) [inline code/game/objects/items/devices/PDA/PDA.dm]: No, this proc logs the message in the server and checks the spamfilters. Actual sending is handled elsewhere.

Comment (PsiOmegaDelta) [inline code/game/objects/items/devices/PDA/PDA.dm]: But you'll still get a rejection message per message server, seeing how the filter is shared by all?
How did you handle the ability to upgrade a server to handle twice the amount of filter entries?

Comment (Kelenius) [inline code/game/objects/items/devices/PDA/PDA.dm]: Hm... you won't get multiple rejection messages, because it returns - but it also means that it will only be logged in the first server.
Updates... ugh. I think I'll just up the limit and remove the upgrading.

**Comment by Minijar** [comment]
Why are we suddenly removing the entirety of telecomms?

**Comment by Zuhayr** [comment]
>'suddenly'
It's been a desirable project for literally a year and a half.

**Comment by Ccomp5950** [comment]
spin up a test box after adding crash_with("MADNESS") to /mob/hear_radio()   say something over the radio and look at what madness telecomms generates.

It's how I spend my saturday mornings :)

File: code/modules/telecomms/telecomms.dm [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	if(!src.on)
+		return 0
+	if(stat & (EMPED|NOPOWER|BROKEN))
+		return 0
+	return 1
+
+/proc/broadcastMessage(var/mob/living/source, var/mobName, var/sayVerb, var/message, var/freq, var/datum/language/speaking, var/messageFlags)
+	var/list/radios = list()
+
+	for(var/obj/item/device/radio/R in world) // AAAAAAAAAAAAAA
```
Comment (Fox-McCloud) [inline code/modules/telecomms/telecomms.dm]: Make a "global_radios" list in radio/new() and ensure it's cleared in radio/destroy()

Also make sure all radios call parent in their own New() and you can avoid this horrible world call.

Comment (Kelenius) [inline code/modules/telecomms/telecomms.dm]: On one hand, thanks for trying to help, on an another, do you really think I don't realize this?

Comment (Lorwp) [inline code/modules/telecomms/telecomms.dm]: On another note....AAAAAAAAAAAAAA ?

**Comment by Kelenius** [comment]
Everything seems to be in working order, but it'd be nice if someone else tested it.

**Comment by Zuhayr** [comment]
I'll pull it down and test it either today or tomorrow.

**Comment by OhEmGee** [comment]
why would you remove this feature

this would mean axing of the tcommsat, a sweet traitor base

**Comment by EmperorJon** [comment]
No, it wouldn't.

**Comment by Kelenius** [comment]
Did anyone test it, is it good to merge?

**Comment by Zuhayr** [comment]
Agh I knew I was forgetting something.

**Comment by Zuhayr** [comment]
Will pull this down and test when I finish labelling all the PRs and issues.

**Comment by Zuhayr** [comment]
Works fine locally. If you can get the conflicts sorted out I'll host it on Europa in the next few days and we can do a proper test by playing a round. Solo local testing is not going to be rigorous enough for this.

**Comment by Zuhayr** [comment]
Testing on byond://server.wetskrell.org:1008 now, will edit bugs into this post as they're found.
- Radios pick up all channels, including encrypted channels. Merc chat seems to be hidden properly.
- Speaker off does not appear to actually stop the headset from passing you radio messages. Hard to tell without the icon.

On another note I really dislike the removal of the radio icon, why did you have to remove it?

**Comment by Kelenius** [comment]
>On another note I really dislike the removal of the radio icon, why did you have to remove it?

Because it's kind of a hassle to maintain here, I'd need to pass the radio as a separate parameter just for that purpose. It wouldn't be hard to code it back in if people want it.

**Comment by Zuhayr** [comment]
I would very much like it. It's good for being able to tell what device is feeding you the message.

**Comment by Kelenius** [comment]
No, it shows the radio used to _send_ the message.

**Comment by Zuhayr** [comment]
I'm an idiot.

File: code/modules/telecomms/telecomms.dm [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+				R.hear_radio(message, sayVerb, speaking, part_a, part_b, source, 1, mobName)
+
+	if(!interceptData && !(freq in ANTAG_FREQS))
+		for(var/antFreq in ANTAG_FREQS)
+			broadcastMessage(source, mobName, sayVerb, message, antFreq, speaking, freq)
+
+	return 1
+
+
+// Round up // REMIND ME TO MOVE THIS TO A BETTER PLACE // Actually nevermind, it's fine
```
Comment (comma) [inline code/modules/telecomms/telecomms.dm]: It's not:
https://github.com/PolarisSS13/Polaris/blob/1e62406b6bce80e2364872bd9ec41ff5b79e9930/code/_helpers/maths.dm#L3
But other things use it so removing it would be better left to another PR.
Not to mention it'll return wrong result on integers (since it always assumes remainder)

**Comment by Kelenius** [comment]
That should do it. Were there any other issues?

**Comment by Kelenius** [comment]
Would it be possible to run a test of this on the actual server?

**Comment by Spookerton** [comment]
If someone can be bothered to make a testing branch and merge this into that, we can certainly run a build of it on live or a different port. Server's plenty beefy enough for five or six side by side with cores to spare. It's mostly a matter of git logistics. *shrug

**Comment by Iamgoofball** [comment]
we removed tcomms at tg, improved pretty much everything

:+1:

**Comment by Datraen** [comment]
Currently has some conflicts, please resolve these and I will try to get something up where the radio can be tested.

**Comment by Kelenius** [comment]
Will probably conflict with #1084, will resolve after merge.

**Comment by Kelenius** [comment]
Resolved, by the way.

**Comment by Kelenius** [comment]
Maybe you could just merge it to playtest it, then revert if it fucks the game up.

**Comment by Kelenius** [comment]
GUYS

**Comment by Zuhayr** [comment]
Why did you close this? The only reason conflicts exist is that you guys are taking 200 years to make a decision about it.
